docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,389 @@
+metadata:
+  name: <name>
+  namespace: <shoot-namespace>
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      name: <label>
+  strategy:
+    inPlaceUpdate:
+      maxUnavailable: 1
+      maxSurge: 0
+      onLabel: false # true/false
+    type: InPlaceUpdate # RollingUpdate/InPlaceUpdate/Recreate
+  template: {}
+    spec: {}
+status: {}
+```
+
+`maxUnavailable`: Specifies the maximum number of unavailable machines during the update process.
+
+`maxSurge`: Specifies the maximum number of machines that can be scheduled above the desired number of machines. Set `maxSurge` to zero if you don't want any new machines/nodes to be provisioned during the update. You can set it to a value greater than zero if you don't want unavailable nodes during the update; in that case, first new machines/nodes will be created with the updated configuration and then the old machines/nodes will undergo in-place updates.
+
+`onLabel`: If the update strategy is `InPlaceUpdateOnLabel` in the worker spec, the worker controller will set this field to true.
+
+### Dependency Watchdog
+
+Currently, the [`dependency-watchdog`](https://github.com/gardener/dependency-watchdog) monitors the leases of the nodes in the shoot cluster. If more nodes than a configured threshold are unhealthy, it scales down the `machine-controller-manager`, `kube-controller-manager`, and `cluster-autoscaler` to prevent a meltdown. During an ongoing in-place update, DWD can make use of the labels added by `machine-controller-manager` during the `in-place` update to exclude nodes undergoing updates from its health check.
+
+### Extensions
+
+#### OS extensions
+
+The OS extension that wants to support in-place updates should add an "update-script" to the `.spec.files[]` of the OSC with a pre-defined path that will be executed by the `gardener-node-agent` to perform the update. This script should trigger the OS-provided update tool/utility to update the OS.
+
+### Failures and Recovery Strategy
+
+In-place updates may fail because of several factors ranging from network issues to compatibility problems. This section outlines the failure scenarios, error classifications, and the specific roles `gardener-node-agent` and `machine-controller-manager` play in failure handling and recovery to ensure minimal disruption.
+
+#### Types of Failures During Update
+
+##### Drain Timeout/Failure
+
+- A node is marked as `ready-for-update` only after a successful drain.
+- We use the existing [drain logic](https://github.com/gardener/machine-controller-manager/tree/v0.54.0/pkg/util/provider/drain) provided by the `machine-controller-manager`.
+- If a drain operation times out or fails, the node will be forcefully drained.
+
+##### Kubernetes Update Failures
+
+- **Rollback on Failure:** If a Kubernetes version update fails, `gardener-node-agent` will roll back the `kubelet` to the version specified in the last applied `OperatingSystemConfig`. This ensures that the node remains operational.
```

Comment:
How will this work? The OSC secret got already updated to the new kubelet config/files, etc. How can it roll back/how will this be orchestrated? Will `gardenlet` update the OSC secret again to the old version? But what happens to the version in the `Shoot` spec then which gets inconsistent?

Reply:
We thought about it and it might not be a very good idea to provide this feature. OS updates needs rollback because we can no longer communicate with the node without a functioning OS. This is not the case with the kubelet. Will update the doc.